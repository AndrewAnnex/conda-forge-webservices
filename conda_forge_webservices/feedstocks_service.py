import git
import jinja2
import os
from .utils import tmp_directory


def handle_feedstock_event(org_name, repo_name):
    if repo_name == "staged-recipes":
        update_listing()
    elif repo_name.endswith("-feedstock"):
        update_feedstock(org_name, repo_name)


def update_listing():
    with tmp_directory() as tmp_dir:
        feedstocks_url = (
            "https://{}@github.com/conda-forge/feedstocks.git"
            "".format(os.environ["FEEDSTOCKS_GH_TOKEN"])
        )
        feedstocks_repo = git.Repo.clone_from(
            feedstocks_url,
            tmp_dir
        )
        feedstocks_dir = os.path.dirname(feedstocks_repo.git_dir)

        repos = sorted(os.listdir(os.path.join(
            feedstocks_dir, "feedstocks"
        )))

        env = jinja2.Environment(
            loader=jinja2.FileSystemLoader(os.path.join(
                os.path.dirname(__file__), "html"
            ))
        )
        context = {"feedstock_names": repos}
        tmpl = env.get_template("feedstocks.html.tmpl")
        feedstocks_html = os.path.join(feedstocks_dir, "feedstocks.html")
        with open(feedstocks_html, 'w') as fh:
            fh.write(tmpl.render(context))

        if feedstocks_repo.is_dirty(untracked_files=True):
            author = git.Actor(
                "conda-forge-coordinator", "conda.forge.coordinator@gmail.com"
            )
            feedstocks_repo.index.add([os.path.relpath(
                feedstocks_html, feedstocks_dir
            )])
            feedstocks_repo.index.commit(
                "Updated the feedstock listing.",
                author=author,
                committer=author
            )
            feedstocks_repo.remote().pull(rebase=True)
            feedstocks_repo.remote().push()


def update_feedstock(org_name, repo_name):
    name = repo_name[:-len("-feedstock")]

    with tmp_directory() as tmp_dir:
        feedstocks_url = (
            "https://{}@github.com/conda-forge/feedstocks.git"
            "".format(os.environ["FEEDSTOCKS_GH_TOKEN"])
        )
        feedstocks_repo = git.Repo.clone_from(
            feedstocks_url,
            tmp_dir
        )

        # Get the submodule
        feedstock_submodule = feedstocks_repo.create_submodule(
            name=name,
            path=os.path.join("feedstocks", name),
            url="https://github.com/{0}/{1}.git".format(org_name, repo_name),
            branch="master"
        )
        # Hack needed if the submodule already exists.
        # Borrows the fix accepted upstream.
        # PR: https://github.com/gitpython-developers/GitPython/pull/679
        feedstock_submodule._name = name

        # Update the feedstocks submodule
        feedstock_submodule.update(init=True, recursive=False, force=True)
        feedstock_submodule.branch.checkout(force=True)
        feedstock_submodule.update(
            init=True,
            recursive=False,
            force=True,
            to_latest_revision=True
        )

        # Submit changes
        if feedstocks_repo.is_dirty(untracked_files=True):
            author = git.Actor(
                "conda-forge-coordinator", "conda.forge.coordinator@gmail.com"
            )
            feedstocks_repo.git.add(update=True)
            feedstocks_repo.index.commit(
                "Updated the {0} feedstock.".format(name),
                author=author,
                committer=author
            )
            feedstocks_repo.remote().pull(rebase=True)
            feedstocks_repo.remote().push()


if __name__ == '__main__':
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('org')
    parser.add_argument('repo')
    args = parser.parse_args()
    update_feedstock(args.org, args.repo)
