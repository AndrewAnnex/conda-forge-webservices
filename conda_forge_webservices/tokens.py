import time
import base64
import os
import io
import sys
import logging
from contextlib import redirect_stdout, redirect_stderr

from github import Github
import jwt
import requests
from cryptography.hazmat.backends import default_backend

LOGGER = logging.getLogger("conda_forge_webservices.tokens")

TOKEN_RESET_TIMES = {}
TEN_MINS = 10*60


def inject_app_token(full_name, repo=None):
    """Inject the webservices app token into the repo secrets.

    Parameters
    ----------
    full_name : str
        The full name of the repo (e.g., "conda-forge/blah").
    repo : pygithub Repository
        Optional repo object to use. If not passed, a new one will be made.

    Returns
    -------
    injected : bool
        True if the token was injected, False otherwise.
    """
    repo_name = full_name.split("/")[1]

    # this is for testing - will turn it on for all repos later
    if repo_name != "cf-autotick-bot-test-package-feedstock":
        return False

    if not repo_name.endswith("-feedstock"):
        return False

    if repo is None:
        gh = Github(os.environ['GH_TOKEN'])
        repo = gh.get_repo(full_name)

    global TOKEN_RESET_TIMES

    now = time.time()
    if TOKEN_RESET_TIMES.get(repo.name, now) <= now + TEN_MINS:
        token = generate_app_token(
            os.environ["CF_WEBSERVICES_APP_ID"],
            os.environ["CF_WEBSERVICES_PRIVATE_KEY"].encode(),
            repo.name,
        )
        if token is not None:
            worked = repo.create_secret("RERENDERING_GITHUB_TOKEN", token)
            if worked:
                TOKEN_RESET_TIMES[repo.name] = Github(token).rate_limiting_resettime
                LOGGER.info("")
                LOGGER.info("===================================================")
                LOGGER.info(
                    "injected app token for repo %s - timeout %ss",
                    repo.name,
                    now - TOKEN_RESET_TIMES[repo.name],
                )
                LOGGER.info("===================================================")
            return worked
        else:
            print("secret creation failed!", flush=True)
            return False
    else:
        return True


def generate_app_token(app_id, raw_pem, repo):
    """Get an app token.

    Parameters
    ----------
    app_id : str
        The github app ID.
    raw_pem : bytes
        An app private key as bytes.
    repo : str
        The name of the repo for which the token is scoped.
        This should be like `ngmix-feedstock` without the org `conda-forge`
        in front.

    Returns
    -------
    gh_token : str
        The github token. May return None if there is an error.
    """
    try:
        if raw_pem[0:1] != b'-':
            raw_pem = base64.b64decode(raw_pem)

        f = io.StringIO()
        with redirect_stdout(f), redirect_stderr(f):
            private_key = default_backend().load_pem_private_key(raw_pem, None)

            ti = int(time.time())
            token = jwt.encode(
                {
                    'iat': ti,
                    'exp': ti + 60*10,
                    'iss': app_id,
                },
                private_key,
                algorithm='RS256',
            )

        if (
            "GITHUB_ACTIONS" in os.environ
            and os.environ["GITHUB_ACTIONS"] == "true"
        ):
            sys.stdout.flush()
            print("masking JWT token for github actions", flush=True)
            print("::add-mask::%s" % token, flush=True)

        with redirect_stdout(f), redirect_stderr(f):
            r = requests.get(
                "https://api.github.com/app/installations",
                headers={
                    'Authorization': 'Bearer %s' % token,
                    'Accept': 'application/vnd.github.machine-man-preview+json',
                },
            )
            r.raise_for_status()

            r = requests.post(
                "https://api.github.com/app/installations/"
                "%s/access_tokens" % r.json()[0]["id"],
                headers={
                    'Authorization': 'Bearer %s' % token,
                    'Accept': 'application/vnd.github.machine-man-preview+json',
                },
                json={"repositories": [repo]},
            )
            r.raise_for_status()

            gh_token = r.json()["token"]

        if (
            "GITHUB_ACTIONS" in os.environ
            and os.environ["GITHUB_ACTIONS"] == "true"
        ):
            sys.stdout.flush()
            print("masking GITHUB token for github actions", flush=True)
            print("::add-mask::%s" % gh_token, flush=True)

        assert r.json()["permissions"] == {
            "contents": "write", "metadata": "read", "workflows": "write",
            "checks": "read", "pull_requests": "write", "statuses": "read",
        }
        assert set(r["name"] for r in r.json()["repositories"]) == {repo}

    except Exception:
        gh_token = None

    return gh_token
