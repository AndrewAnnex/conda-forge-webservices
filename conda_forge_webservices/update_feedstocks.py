import git
import github
import os
from .utils import tmp_directory


def update_feedstock(org_name, repo_name):
    if not repo_name.endswith("-feedstock"):
        return

    gh = github.Github(os.environ['FEEDSTOCKS_GH_TOKEN'])

    repo_gh = gh.get_repo("{}/{}".format(org_name, repo_name))
    name = repo_name[:-len("-feedstock")]

    with tmp_directory() as tmp_dir:
        feedstocks_url = (
            "https://{}@github.com/conda-forge/feedstocks.git"
            "".format(os.environ["FEEDSTOCKS_GH_TOKEN"])
        )
        feedstocks_repo = git.Repo.clone_from(
            feedstocks_url,
            tmp_dir
        )

        # Get the submodule
        feedstock_submodule = feedstocks_repo.create_submodule(
            name=name,
            path=os.path.join("feedstocks", name),
            url=repo_gh.clone_url,
            branch="master"
        )
        # Hack needed if the submodule already exists.
        # Borrows the fix accepted upstream.
        # PR: https://github.com/gitpython-developers/GitPython/pull/679
        feedstock_submodule._name = name

        # Update the feedstocks submodule
        feedstock_submodule.update(init=True, recursive=False, force=True)
        feedstock_submodule.branch.checkout(force=True)
        feedstock_submodule.update(
            init=True,
            recursive=False,
            force=True,
            to_latest_revision=True
        )

        # Submit changes
        if feedstocks_repo.is_dirty():
            author = git.Actor(
                "conda-forge-coordinator", "conda.forge.coordinator@gmail.com"
            )
            feedstocks_repo.git.add(update=True)
            feedstocks_repo.index.commit(
                "Updated feedstocks submodules. [ci skip]",
                author=author,
                committer=author
            )
            feedstocks_repo.remote().pull(rebase=True)
            feedstocks_repo.remote().push()


if __name__ == '__main__':
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('org')
    parser.add_argument('repo')
    args = parser.parse_args()
    update_team(args.org, args.repo)
