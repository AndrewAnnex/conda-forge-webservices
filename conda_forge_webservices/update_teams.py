from git import GitCommandError, Repo
from conda_build.metadata import MetaData
import github
import os

from .utils import tmp_directory

def update_team(org_name, repo_name):
    if not repo_name.endswith("-feedstock"):
        return

    gh = github.Github(os.environ['GH_TOKEN'])
    org = gh.get_organization(org_name)
    remote_repo = org.get_repo(repo_name)

    with tmp_directory() as tmp_dir:
        repo = Repo.clone_from(remote_repo.clone_url, tmp_dir)
        meta = MetaData(tmp_dir)
        updated_maintainers = set(meta.meta.get('extra', {}).get('recipe-maintainers', []))

        team_name = repo_name[:-len("-feedstock")]
        team = next(team for team in org.get_teams() if team.name == team_name)
        current_maintainers = set([maintainer.login.lower() for maintainer in team.get_members()])

        for new_maintainer in updated_maintainers - current_maintainers:
            headers, data = team._requester.requestJsonAndCheck(
                "PUT",
                team.url + "/memberships/" + new_maintainer
            )

        for old_maintainer in current_maintainers - updated_maintainers:
            headers, data = team._requester.requestJsonAndCheck(
                "DELETE",
                team.url + "/memberships/" + old_maintainer
            )


if __name__ == '__main__':
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('org')
    parser.add_argument('repo')
    args = parser.parse_args()
    update_team(args.org, args.repo)
