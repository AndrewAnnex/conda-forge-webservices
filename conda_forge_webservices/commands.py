from git import GitCommandError, Repo, Actor
from conda_build.metadata import MetaData
import github
import os
import subprocess
from .utils import tmp_directory
from .linting import compute_lint_message, comment_on_pr, set_pr_status
from .update_teams import update_team
from .circle_ci import update_circle
from conda_smithy import __version__ as conda_smithy_version
import textwrap

ADD_NOARCH_MSG = "please add noarch: python"
RERENDER_MSG = "please rerender"
LINT_MSG = "please lint"
UPDATE_TEAM_MSG = "please update team"
UPDATE_CIRCLECI_KEY_MSG = "please update circle"

def check_for_bot(comment):
    return "@conda-forge-admin" in comment or "@conda-forge-linter" in comment


def pr_comment(org_name, repo_name, issue_num, comment):
    if not check_for_bot(comment):
        return
    gh = github.Github(os.environ['GH_TOKEN'])
    repo = gh.get_repo("{}/{}".format(org_name, repo_name))
    pr = repo.get_pull(int(issue_num))
    pr_detailed_comment(org_name, repo_name, pr.head.user.login, pr.head.repo.name, pr.head.ref, issue_num, comment)


def pr_detailed_comment(org_name, repo_name, pr_owner, pr_repo, pr_branch, pr_num, comment):
    is_staged_recipes = (repo_name == "staged-recipes")
    if not (repo_name.endswith("-feedstock") or is_staged_recipes):
        return

    if not check_for_bot(comment):
        return

    pr_commands = [LINT_MSG]
    if not is_staged_recipes:
        pr_commands += [ADD_NOARCH_MSG, RERENDER_MSG]

    if not any(command in comment.lower() for command in pr_commands):
        return

    with tmp_directory() as tmp_dir:
        feedstock_dir = os.path.join(tmp_dir, repo_name)
        repo_url = "https://{}@github.com/{}/{}.git".format(os.environ['GH_TOKEN'],
            pr_owner, pr_repo)
        repo = Repo.clone_from(repo_url, feedstock_dir, branch=pr_branch)

        if ADD_NOARCH_MSG in comment.lower():
            make_noarch(repo)
            rerender(repo, pr_num)
        if RERENDER_MSG in comment.lower():
            rerender(repo, pr_num)
        if LINT_MSG in comment.lower():
            relint(org_name, repo_name, pr_num)

        repo.remotes.origin.push()


def issue_comment(org_name, repo_name, issue_num, title, comment):
    if not repo_name.endswith("-feedstock"):
        return

    comment = comment.lower()
    title = title.lower()

    if not check_for_bot(comment + title):
        return

    issue_commands = [UPDATE_TEAM_MSG, ADD_NOARCH_MSG, UPDATE_CIRCLECI_KEY_MSG]

    if not any(command in (comment+title) for command in issue_commands):
        return

    gh = github.Github(os.environ['GH_TOKEN'])
    repo = gh.get_repo("{}/{}".format(org_name, repo_name))
    issue = repo.get_issue(int(issue_num))

    if UPDATE_TEAM_MSG in comment + title:
        update_team(org_name, repo_name)
        if UPDATE_TEAM_MSG in title:
            issue.edit(state="closed")
        message = textwrap.dedent("""
                Hi! This is the friendly automated conda-forge-webservice.

                I just wanted to let you know that I updated the team with maintainers from master.
                """)
        issue.create_comment(message)

    if UPDATE_CIRCLECI_KEY_MSG in comment + title:
        update_circle(org_name, repo_name)
        if UPDATE_CIRCLECI_KEY_MSG in title:
            issue.edit(state="closed")
        message = textwrap.dedent("""
                Hi! This is the friendly automated conda-forge-webservice.

                I just wanted to let you know that I updated the circle-ci deploy key and followed the project.
                """)
        issue.create_comment(message)

    forked_user = gh.get_user().login
    forked_repo = gh.get_user().create_fork(repo)

    with tmp_directory() as tmp_dir:
        feedstock_dir = os.path.join(tmp_dir, repo_name)
        repo_url = "https://{}@github.com/{}/{}.git".format(os.environ['GH_TOKEN'],
            forked_user, repo_name)
        git_repo = Repo.clone_from(repo_url, feedstock_dir)
        forked_repo_branch = 'conda_forge_admin_{}'.format(issue_num)
        new_branch = git_repo.create_head(forked_repo_branch)
        new_branch.checkout()

        if ADD_NOARCH_MSG in comment + title:
            make_noarch(git_repo)
            rerender(git_repo)
            git_repo.git.push("origin", forked_repo_branch)
            msg = "MNT: Add noarch: python"
            pr = repo.create_pull(msg, "As instructed in #{}".format(issue_num),
                    "master", "{}:{}".format(forked_user, forked_repo_branch))

            if ADD_NOARCH_MSG in title:
                issue.edit(state="closed")

            message = textwrap.dedent("""
                    Hi! This is the friendly automated conda-forge-webservice.

                    I just wanted to let you know that I made the recipe noarch: python in {}/{}#{}.
                    """.format(org_name, repo_name, pr.number))
            issue.create_comment(message)


def rerender(repo, pr_num):
    subprocess.call(["conda", "smithy", "rerender"], cwd=repo.working_dir)
    if repo.is_dirty():
        author = Actor("conda-forge-admin", "pelson.pub+conda-forge@gmail.com")
        repo.index.commit("MNT: Re-rendered with conda-smithy {}".format(conda_smithy_version), author=author, committer=author)
    else:
        message = textwrap.dedent("""
                Hi! This is the friendly automated conda-forge-webservice.

                I rerendered the feedstock and it seems to be already up-to-date.
                """)
        repo.get_issue(pr_num).create_comment(message)


def make_noarch(repo):
    meta_yaml = os.path.join(repo.working_dir, "recipe", "meta.yaml")
    with open(meta_yaml, 'r') as fh:
        lines = [line for line in fh]
    with open(meta_yaml, 'w') as fh:
        build_line = False
        for line in lines:
            if build_line:
                spaces = len(line) - len(line.lstrip())
                line = "{}noarch: python\n{}".format(" "*spaces, line)
            build_line = False
            if line.rstrip() == 'build:':
                build_line = True
            fh.write(line)
    repo.index.add([meta_yaml])
    author = Actor("conda-forge-admin", "pelson.pub+conda-forge@gmail.com")
    repo.index.commit("Add noarch:python option", author=author)


def relint(owner, repo_name, pr_num):
    pr = int(pr_num)
    lint_info = compute_lint_message(owner, repo_name, pr, False)
    if not lint_info:
        print('Linting was skipped.')
    else:
        msg = comment_on_pr(owner, repo_name, pr, lint_info['message'])
        set_pr_status(owner, repo_name, lint_info, target_url=msg.html_url)

