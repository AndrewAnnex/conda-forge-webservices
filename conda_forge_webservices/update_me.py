import os
import subprocess
import json
import tempfile
import shutil
import logging

from git import Repo
import requests

from conda_forge_webservices.tokens import get_app_token_for_webservices_only

# from .utils import tmp_directory

LOGGER = logging.getLogger("conda_forge_webservices.update_me")

PKGS = ["conda-smithy"]


def _run_git_command(args):
    subprocess.run(['git'] + args, check=True)


def get_current_versions():
    r = subprocess.run(["conda", "list"], capture_output=True)
    out = r.stdout.decode("utf-8")
    vers = {}
    for line in out.split("\n"):
        for pkg in PKGS:
            if pkg in line:
                items = line.split()
                vers[pkg] = items[1]
    return vers


def main():
    """Get current versions from the heroku app and update if they are old.

    Note this script runs on GHA, not on the heroku app.
    """
    # keep these imports here to protect the webservice from memory errors
    # due to conda
    from conda.core.index import get_index
    from conda.models.match_spec import MatchSpec
    from conda.models.version import VersionOrder
    from conda.resolve import Resolve

    r = requests.get(
        "https://conda-forge.herokuapp.com/conda-webservice-update/versions")
    r.raise_for_status()
    installed_vers = r.json()

    index = get_index(channel_urls=['conda-forge'])
    r = Resolve(index)

    to_install = {}
    final_install = {}

    for pkg in PKGS:
        available_versions = [
            p.version for p in r.get_pkgs(MatchSpec(pkg))
            if "conda-forge" in str(p.channel)
        ]
        available_versions = sorted(available_versions, key=VersionOrder)
        latest_version = available_versions[-1]
        LOGGER.info("%s|latest|installed:" % pkg, latest_version, installed_vers[pkg])
        if VersionOrder(latest_version) != VersionOrder(installed_vers[pkg]):
            to_install[pkg] = latest_version
            final_install[pkg] = latest_version
        else:
            final_install[pkg] = installed_vers[pkg]

    if to_install:
        tmpdir = None
        try:
            gh_token = get_app_token_for_webservices_only()
            tmpdir = tempfile.mkdtemp('_recipe')

            repo_name = "conda-forge-webservices"
            clone_dir = os.path.join(tmpdir, repo_name)
            url = "https://x-access-token:{}@github.com/conda-forge/{}.git".format(
                gh_token, repo_name
            )

            repo = Repo.clone_from(url, clone_dir, depth=1)

            # keep a record around
            pth = os.path.join(clone_dir, "pkg_versions.json")
            with open(pth, "w") as fp:
                json.dump(final_install, fp)
            repo.index.add(pth)

            msg_vers = ", ".join(["{}={}".format(k, v) for k, v in to_install.items()])
            repo.index.commit("redeploy for '%s'" % msg_vers)
            repo.git.push("origin", "main")

        finally:
            if tmpdir is not None:
                shutil.rmtree(tmpdir)
