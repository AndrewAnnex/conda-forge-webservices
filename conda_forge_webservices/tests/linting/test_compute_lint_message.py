from __future__ import print_function
from collections import OrderedDict
from contextlib import contextmanager
import os
import shutil
import subprocess
import tempfile
import textwrap
import unittest

from conda_forge_webservices.linting import compute_lint_message


@contextmanager
def tmp_directory():
    tmp_dir = tempfile.mkdtemp('recipe_')
    yield tmp_dir
    shutil.rmtree(tmp_dir)


class Test_compute_lint_message(unittest.TestCase):
    def test_good_recipe(self):
        expected_message = textwrap.dedent("""
        Hi! This is the friendly conda-forge-admin automated user.
        
        I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/good_recipe```) and found it was in an excellent condition.

        """)

        lint = compute_lint_message('conda-forge', 'conda-forge-webservices', 16)
        self.assertMultiLineEqual(expected_message, lint['message'])

    def test_bad_recipe(self):
        expected_message = textwrap.dedent("""
        Hi! This is the friendly conda-forge-admin automated user.
        
        I wanted to let you know that I linted all conda-recipes in your PR (```recipes/bad_recipe```) and found some lint.
        
        Here's what I've got...
        
        
        For **recipes/bad_recipe**:
        
         * The home item is expected in the about section.
         * The license item is expected in the about section.
         * The summary item is expected in the about section.
         * The recipe could do with some maintainers listed in the "extra/recipe-maintainers" section.
         * The recipe must have some tests.
        """)

        lint = compute_lint_message('conda-forge', 'conda-forge-webservices', 17)
        self.assertMultiLineEqual(expected_message, lint['message'])

    def test_no_recipe(self):
        expected_message = textwrap.dedent("""
        Hi! This is the friendly conda-forge-admin automated user.

        I was trying to look for recipes to lint for you, but couldn't find any.
        Please ping the 'conda-forge/core' team (using the @ notation in a comment) if you believe this is a bug.
        """)

        lint = compute_lint_message('conda-forge', 'conda-forge-webservices', 18)
        self.assertMultiLineEqual(expected_message, lint['message'])


if __name__ == '__main__':
    unittest.main()
