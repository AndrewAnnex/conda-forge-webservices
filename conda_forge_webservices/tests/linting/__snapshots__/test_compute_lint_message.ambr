# serializer version: 1
# name: test_new_recipe[rattler-build-good-lint]
  tuple(
    '''
      
      Hi! This is the friendly automated conda-forge-linting service.
      
      I wanted to let you know that I linted all conda-recipes in your PR (```meta.yaml```) and found some lint.
      
      Here's what I've got...
      
      
      For **meta.yaml**:
      
       * Failed to even lint the recipe, probably because of a conda-smithy bug :cry:. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint .` from the recipe directory. 
  
    ''',
    'bad',
  )
# ---
